packages/docs/docs/getting-started/create-unirep-app.md

```diff
@@ -159,114 +142,129 @@ Requesting data with epoch key:
+If the proof is invalid, the snarkjs prover will throw an error.
 
+**e.g.**<br/>
 If user's provable data is:
 ```shell
 Provable Data 0 = 2 👈
 Provable Data 1 = 3
 Provable Data 2 = 4
 Provable Data 3 = 5
 ```
-and user claims he has data:
+and they claim to own the data:
 ```shell
 Claim Data 0 = 3 👈
 Claim Data 1 = 3
 Claim Data 2 = 3
 Claim Data 3 = 3
 ```
 
-It throws an error since the fields where the emojis are do not satisfy the conditions:
+The prover will throw an error because the fields with emojis do not satisfy the conditions:
 
 ```shell
 2 (Provable Data 0) >= 3 (Claim Data 0) # not satisfied. 2 is not greater than or equal to 3
 ```
 
+:::info
+See the customized [data proof circuit](https://github.com/Unirep/create-unirep-app/blob/2ca9aaa3fcacb9282993b8f5d5917afc482ec089/packages/circuits/circuits/dataProof.circom)<br/>
+App builders can customize their own ZK circuits and deploy verifiers to fit the application.
+:::
+
 ## 💡 Build your own application
 
-Start with modifying the [`UnirepApp.sol`](https://github.com/Unirep/create-unirep-app/blob/2ca9aaa3fcacb9282993b8f5d5917afc482ec089/packages/contracts/contracts/UnirepApp.sol).
+Start by modifying the [`UnirepApp.sol`](https://github.com/Unirep/create-unirep-app/blob/2ca9aaa3fcacb9282993b8f5d5917afc482ec089/packages/contracts/contracts/UnirepApp.sol) contract.
 
 :::info
 See all [`Unirep.sol` APIs](../contracts-api/unirep-sol.md)
 :::
 
-### Change epoch length
+### Sign up users
+
+Attesters may allow anyone to join and use their application, as in the example attester, or..
+
+e.g:
+
+<!-- TODO:
+provide ideas how an attester can require zk proof for users to join, example from zketh? -->
+### Define epoch length
+
+Each attester sets its own [epoch](../protocol/epoch.md) length, determined by the needs of the application. When deciding their attester's epoch length, developers should consider the following trade-offs.<br/>
 
-Each attester has its own [epoch](../protocol/epoch.md) length. <br/>
-If the epoch is too long: users will use the same epoch keys for a long time, new data cannot be proved immediately, and you risk making users pseudoanonymous because of the re-use of epoch keys. <br/>
-If the epoch is too short: epoch keys last for a short period and the user does not have enough time to receive much data.
+longer epoch lengths:
+- allow users more time to receive new data, especially in cases with user-to-user interactions
+- extend the time users must wait for changed data to be provable
+- risk making users pseudoanonymous, as their activities may be tracked when using the same epoch keys for an long period of time 
+
+shorter epoch lengths: 
+- allow users to prove new data sooner
+- limit opportunities for users to receive new data, as their epoch keys expire quickly
+- may enable a higher degree of user anonymity, as user's anonymous identifiers change more frequently
```

Comment:
this section feels like a good way to help devs grasp protocol functionality, is there anything else we could add here, considerations that might help them with design decisions in the ideation phase?

Reply:
yeah it looks better 👍🏻